### PR TITLE
Add job submission script of Cori Haswell in Doc

### PR DESCRIPTION
**Pull request by Yin-YinjianZhao**
This PR add some information to help users running jobs on Cori Haswell with good performance.

### CODE REVIEW

File: Docs/source/running_cpp/platforms.rst [hunk 1]
```diff
@@ -37,6 +37,29 @@ regime), the following set of parameters provided good performance:
 
 * **2 grids per MPI**, *i.e.*, 16 grids per KNL node.
 
+Running on Cori Haswell at NERSC
+----------------------------
```
Comment (ax3l) [inline Docs/source/running_cpp/platforms.rst]: Little `.rst` quirk:
```suggestion
Running on Cori Haswell at NERSC
--------------------------------
```

File: Docs/source/running_cpp/platforms.rst [hunk 2]
```diff
@@ -37,6 +37,29 @@ regime), the following set of parameters provided good performance:
...
+  sbatch batch_cori_haswell.sh
+
+to submit the job.
+
+For a 3D simulation with a few (1-4) particles per cell using FDTD Maxwell
+solver on Cori Haswell for a well load-balanced problem (in our case laser
+wakefield acceleration simulation in a boosted frame in the quasi-linear
+regime), the following set of parameters provided good performance:
+
+* **4 MPI ranks per Haswell node**, with ``OMP_NUM_THREADS=16``
```
Comment (ax3l) [inline Docs/source/running_cpp/platforms.rst]:
```suggestion
* **4 MPI ranks per Haswell node** (2 MPI ranks per `Intel Xeon E5-2698 v3 <https://ark.intel.com/content/www/us/en/ark/products/81060/intel-xeon-processor-e5-2698-v3-40m-cache-2-30-ghz.html>`_), with ``OMP_NUM_THREADS=16`` (which uses `2x hyperthreading <https://docs.nersc.gov/jobs/affinity/>`_)
```

File: Docs/source/running_cpp/platforms.rst [hunk 3]
```diff
@@ -37,6 +37,29 @@ regime), the following set of parameters provided good performance:
 
 * **2 grids per MPI**, *i.e.*, 16 grids per KNL node.
 
+Running on Cori Haswell at NERSC
+----------------------------
+
+The batch script below can be used to run a WarpX simulation on 1 Haswell node on
+the supercomputer Cori at NERSC.
```
Comment (ax3l) [inline Docs/source/running_cpp/platforms.rst]:
```suggestion
The batch script below can be used to run a WarpX simulation on 1 `Haswell node <https://docs.nersc.gov/systems/cori/>`_ on the supercomputer Cori at NERSC.
```

File: Tools/BatchScripts/batch_cori_haswell.sh [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+#SBATCH -e error.txt
+#SBATCH -o output.txt
+#SBATCH --tasks-per-node=4
+#SBATCH --cpus-per-task=16
+
+export OMP_PROC_BIND=spread
+export OMP_PLACES=threads
+export OMP_NUM_THREADS=16
+
+export WARPX_NMPI_PER_NODE=4
```
Comment (ax3l) [inline Tools/BatchScripts/batch_cori_haswell.sh]:
```suggestion
# each Cori Haswell node has 2 sockets of Intel Xeon E5-2698 v3
# each Xeon CPU is divided into 2 bus rings that each have direct L3 access
export WARPX_NMPI_PER_NODE=4
```

File: Tools/BatchScripts/batch_cori_haswell.sh [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+#SBATCH -J <job name>
+#SBATCH -A <allocation ID>
+#SBATCH -e error.txt
+#SBATCH -o output.txt
+#SBATCH --tasks-per-node=4
+#SBATCH --cpus-per-task=16
+
+export OMP_PROC_BIND=spread
+export OMP_PLACES=threads
+export OMP_NUM_THREADS=16
```
Comment (ax3l) [inline Tools/BatchScripts/batch_cori_haswell.sh:32]: Easier to explain if set below:
```suggestion
```

File: Tools/BatchScripts/batch_cori_haswell.sh [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+export WARPX_NMPI_PER_NODE=4
+
```
Comment (ax3l) [inline Tools/BatchScripts/batch_cori_haswell.sh]:
```suggestion

# each MPI rank per half-socket has 8 physical cores
#   or 16 logical (virtual) cores
# over-subscribing each physical core with 2x
#   hyperthreading leads to a slight (3.5%) speedup
# the settings below make sure threads are close to the
#   controlling MPI rank (process) per half socket and
#   distribute equally over close-by physical cores and,
#   for N>8, also equally over close-by logical cores
export OMP_PROC_BIND=spread
export OMP_PLACES=threads
export OMP_NUM_THREADS=16

```

File: Tools/BatchScripts/batch_cori_haswell.sh [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+#SBATCH -N 1
+#SBATCH -t 03:00:00
+#SBATCH -q regular
+#SBATCH -C haswell
+#SBATCH -J <job name>
+#SBATCH -A <allocation ID>
+#SBATCH -e error.txt
+#SBATCH -o output.txt
+#SBATCH --tasks-per-node=4
+#SBATCH --cpus-per-task=16
```
Comment (ax3l) [inline Tools/BatchScripts/batch_cori_haswell.sh]:
```suggestion
# one MPI rank per half-socket (see below)
#SBATCH --tasks-per-node=4
# request all logical (virtual) cores per half-socket
#SBATCH --cpus-per-task=16
```

File: Tools/BatchScripts/batch_cori_haswell.sh [hunk 8]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash -l
+
+#SBATCH -N 1
```
Comment (ax3l) [inline Tools/BatchScripts/batch_cori_haswell.sh:4]:
```suggestion
# Just increase this number of you need more nodes.
#SBATCH -N 1
```

**Comment by ax3l** [comment]
Just for documentation:

We documented data scans in our google drive under `WarpX > Topics > CodeComparisons > EPOCH > 2020_LLNL_performance_tuning > WarpX_EPOCH profiling`.

Resources we used (the important ones are in the PR, too):

- https://docs.nersc.gov/systems/cori/
- https://en.wikichip.org/wiki/intel/xeon_e5
- https://en.wikichip.org/wiki/intel/microarchitectures/haswell
- https://www.enterpriseai.news/2014/09/08/intel-ups-performance-ante-haswell-xeon-chips/ (See the `14-18 Core (HCC)` layout for the dual-bus layout with L3 access for the Intel Xeon Processor E5-2698 v3)

**Review by ax3l — APPROVED** [review]
Thank you for your help figuring out the optimal parameters! :sparkles:

**Review by mrowan137 — APPROVED** [review]

**Pull request merged**